okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt

```diff
@@ -82,6 +109,13 @@ class OkHttpClientTestRule : TestRule {
      */
     internal val prototypes = ConcurrentLinkedDeque<OkHttpClient>()
 
+    val clientEventsMap = mutableMapOf<OkHttpClient, MutableList<String>>()
```

Comment:
Why isn’t this just a property of the OkHttpClientTestRule?

Reply:
Because of static caching of the client instances, doesn't match the Rule lifecycle.

Reply:
Got it. Lets put the current `Rule` in the `EventListener.Factory`, so each `EventListener` knows which `Rule` created it.

Reply:
Maybe something like this:

```
  class OkHttpClientTestRule {
    val lines = mutableListOf<String>()

    ...

    private fun freshClient(): OkHttpClient {
      return OkHttpClient.Builder()
          .dns(SINGLE_INET_ADDRESS_DNS) // Prevent unexpected fallback addresses.
          .eventListenerFactory(object: EventListener.Factory {
             fun create(call: Call) = ClientRuleEventListener { client, line ->
               this@OkHttpClientTestRule.events.add(line)
             }
         })
         .build()
```

Reply:
Nice, think I finally grokked it.